libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+  /// Should only be used to handle multiple active spans simultaneously
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, const ProgressContext& child_of) = 0;
+
+  /// Calls finish on the active span, its parent is set to be active
+  /// Primarily for internal use only, preferred to use ProgressSpan's Finish function
+  void FinishActiveSpan();
+
+  // For passing spans across process/host boundaries
+  // These functions are needed, but are implemented now for
+  // debugging purposes, they will be replaced
+  // Extract returns a nullptr on failure
+  virtual std::string Inject(const ProgressContext& ctx) = 0;
+  virtual std::unique_ptr<ProgressContext> Extract(
+      const std::string& carrier) = 0;
+
+  /// Get the current scope’s span to add tagging/logging without having
+  /// to pass around scopes/spans as parameters
+  /// If there is no active span, an unnamed root span of a new trace
+  /// is created (in this case the program is probably not using tracing)
+  virtual ProgressSpan& GetActiveSpan();
+  bool HasActiveSpan() { return active_span_ != nullptr; }
+  uint32_t GetHostID() { return host_id_; }
+  uint32_t GetNumHosts() { return num_hosts_; }
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
+  virtual void Close() = 0;
+
+private:
+  static std::unique_ptr<ProgressTracer> tracer_;
+
+  std::shared_ptr<ProgressSpan> active_span_ = nullptr;
+  uint32_t host_id_;
+  uint32_t num_hosts_;
+  std::shared_ptr<ProgressSpan> default_active_span_ = nullptr;
+};
+
+class KATANA_EXPORT ProgressScope {
+  friend class ProgressTracer;
+
+  ProgressScope(std::shared_ptr<ProgressSpan> span) : span_(span) {}
+
+public:
+  ~ProgressScope();
+  ProgressScope(const ProgressScope&) = delete;
+  ProgressScope(ProgressScope&&) = delete;
+  ProgressScope& operator=(const ProgressScope&) = delete;
+  ProgressScope& operator=(ProgressScope&&) = delete;
+
+  /// Get the scope’s underlying span to add tagging/logging,
+  /// span relationships, and handle multiple active spans at a time
+  ProgressSpan& span() { return *span_; }
+
+  /// Closes the underlying ProgressSpan if the ProgressScope was created
+  /// with the flag finish_on_close=true
+  /// Note that this will only close the underlying ProgressSpan when all
+  /// of its active children ProgressSpans have been finished
+  /// This will be called by RAII if it has not already been called
+  void Close();
+
+private:
+  std::shared_ptr<ProgressSpan> span_;
+};
+
+class KATANA_EXPORT ProgressContext {
+public:
+  virtual ~ProgressContext() = default;
+  virtual std::unique_ptr<ProgressContext> Clone() const noexcept = 0;
+  virtual std::string GetTraceID() const noexcept;
+  virtual std::string GetSpanID() const noexcept;
+};
+
+class KATANA_EXPORT ProgressSpan {
+  virtual void Close() = 0;
```

Comment:
This is a pure virtual private method and JSONSpan actually makes this public. Should this be at least protected?

Reply:
That's a bug actually, thanks for spotting it, I'll make it private